### ISSUE TITLE
Interruptible eval *msg* and middlewares

### ISSUE THREAD

**Issue by jpmonettas** [nrepl/nrepl#363]
Since 1.3.0-beta3 [the session is binding](https://github.com/nrepl/nrepl/commit/483d32b8370a5e9fa51c1cbadc00047b139f87e2#diff-be175d655029beb7840642730495d2ce791cb10f156f32222217119206271009R353) `*msg*` to the original message. 
This seams to break middlewares that as part of the chain are for example wrapping the transport in order to look at stuff being written to `*out*` or `*err*` I think because session-exec is now using `*msg*`  

Since middlewares down the road (or handlers) now can choose to work with `msg` or `*msg*` (which could be different) as a middleware author I guess you now need to remember to also re-bind `*msg*` once you modified it? possibly breaking other assumptions?  

### Clojure version

All versions

### Java version

All versions

### Operating system

All OSs

**Comment by alexander-yakushev** [comment]
Sorry, still don't quite understand the scenario which this change breaks. Could you provide an approximate example?

**Comment by jpmonettas** [comment]
@alexander-yakushev Sure. So I had a middleware that in order to intercept stuff written to `*out*` and `*err*` was doing something like this (stripped to a minimal) : 

```clojure
(defn wrap-middleware
  [next-handler]
  (fn [msg]
    (next-handler (update msg
                          :transport
                          (fn [transport]
                            (reify Transport
                              (recv [_this]
                                (t/recv transport))
                              (recv [_this timeout]
                                (t/recv transport timeout))
                              (send [this resp]
                                (cond
                                  (:out resp) (storing-out-somewhere (:out resp))
                                  (:err resp) (storing-err-somewhere (:err resp)))
                                (.send transport resp)
                                this)))))))
```

So basically somewhere in the chain wrapping the transport so whatever down the chain writes to `*out*` or `*err*` the middleware can capture it.
This was working until beta3. In beta3 the session middleware is now binding the input message (the one with the original transport) to interruptible eval `*msg*`, which is the message being used by the eval code.

I was able to fix my middleware by re-binding `*msg*` after updating it with my new wrapped transport, but I think this is brittle.

The problem I see is that middlewares can update messages as the move through the chain, but then eval could be working with whatever is on `*msg*`, which could be different that the `msg` that got to eval.

Does it make sense?

**Comment by jpmonettas** [comment]
In case it wasn't clear, this is what I'm doing now :
```clojure
(defn wrap-flow-storm
  [next-handler]
  (fn [msg]
    (let [msg' (update msg
                       :transport
                       (fn [transport]
                         (reify Transport
                           (recv [_this]
                             (t/recv transport))
                           (recv [_this timeout]
                             (t/recv transport timeout))
                           (send [this resp]
                             (cond
                               (:out resp) (storing-out-somewhere (:out resp))
                               (:err resp) (storing-err-somewhere (:err resp)))
                             (.send transport resp)
                             this))))])
    (binding [*msg* msg']
      (next-handler msg'))))
```

**Comment by alexander-yakushev** [comment]
Yes, right, I get it now. It is a problem indeed. Guess I'm never gonna be happy with the solution here, but I'll revisit it once more.

**Comment by alexander-yakushev** [comment]
@jpmonettas We released the fix in #363 as nREPL `1.3.1`. Could you please check if it solves your problem?

**Comment by jpmonettas** [comment]
It works now! Thanks @alexander-yakushev
